File: backend/schemas/exam.py
from enum import Enum
from typing import Any, Dict, List, Optional

from pydantic import BaseModel


class ExamType(str, Enum):
    PRE_EXAM = "pre_exam"
    QUIZ = "quiz"
    FINAL_EXAM = "final_exam"


# without is_correct
class OptionBase(BaseModel):
    id: Optional[int] = None
    option_text: str


# with is_correct
class Option(OptionBase):
    is_correct: bool


# in create exam request
class PostQuestionBase(BaseModel):
    question_text: str
    options: List[Option]


class GetQuestionBase(BaseModel):
    question_text: str
    options: List[OptionBase]


# in get exam response
class Question(GetQuestionBase):
    id: int
    exam_id: int


class PostExamRequest(BaseModel):
    module_id: Optional[int] = None
    course_id: Optional[int] = None
    questions: List[PostQuestionBase]


class QuestionResponse(BaseModel):
    id: int
    exam_id: int
    question_text: str
    options: List[Option]


class PostExamResponse(BaseModel):
    exam_id: int
    questions: List[QuestionResponse]


class GetExamResponse(BaseModel):
    id: int
    module_id: Optional[int]
    course_id: Optional[int]
    questions: List[Question]


class ExamSubmissionRequest(BaseModel):
    exam_id: int
    exam_type: ExamType
    answers: List[Dict[str, Any]]  # [{"question_id": 1, "selected_option_id": 3}, ...]


class ExamSubmissionResponse(BaseModel):
    success: bool
    score: int
    total_questions: int
    correct_answers: int
    passed: bool
    progress_updated: bool
    message: str
    next_available_lesson_id: int | None = None
    next_available_module_id: int | None = None
    next_available_exam_id: int | None = None

File: schemas/Exam.ts
export enum ExamType {
  PRE_EXAM = "pre_exam",
  QUIZ = "quiz",
  FINAL_EXAM = "final_exam",
  ACTIVITY = "activity",
}

export interface Exam {
  id: number;
  module_id: number | null;
  course_id: number | null;
  time: number;
  questions: Question[];
}

export interface Question {
  id: number;
  exam_id: number;
  question_text: string;
  options: Option[];
}

export interface Option {
  id: number;
  option_text: string;
}

export interface ExamAnswer {
  question_id: number;
  selected_option_id: number;
}

export interface ExamSubmissionRequestBody {
  exam_id: number;
  exam_type: ExamType;
  answers: ExamAnswer[];
}

export interface ExamSubmissionResult {
  success: boolean;
  score: number;
  total_questions: number;
  correct_answers: number;
  passed: boolean;
  message: string;
  next_available_exam_id: number | null;
  next_available_lesson_id: number | null;
  next_available_module_id: number | null;
  progress_updated: boolean;
  detailed_review?: QuestionReview[] | null;
}

export interface OptionReview extends Option {
  is_correct: boolean;
}

export interface QuestionReview extends Question {
  options: OptionReview[]; // Use the new extended OptionReview interface
  submitted_option_id: number | null;
  is_correctly_answered: boolean;
}

export interface AlreadySubmittedResponse {
  status: "already_submitted";
  error: string;
  exam_type: string;
  result: ExamSubmissionResult; // The actual submission data
}
